payjoin/src/send/error.rs

```diff
@@ -235,3 +235,191 @@ impl std::error::Error for CreateRequestError {
+                },
+            Err(e) => return e.into(),
+        }
+    }
+    /// Parse a response from the receiver.
+    ///
+    /// response must be valid JSON string.
+    pub fn from_str(response: &str) -> Self {
+        if let Some(parsed) = serde_json::from_str(response).ok() {
+            Self::from_json(parsed)
+        } else {
+            InternalValidationError::Parse.into()
+        }
+    }
+}
+
+impl std::error::Error for ResponseError {}
+
+impl From<WellKnownError> for ResponseError {
+    fn from(value: WellKnownError) -> Self { Self::WellKnown(value) }
+}
+
+impl From<InternalValidationError> for ResponseError {
+    fn from(value: InternalValidationError) -> Self {
+        Self::Validation(ValidationError { internal: value })
+    }
+}
+
+// It is imperative to carefully display pre-defined messages to end users and the details in debug.
+impl Display for ResponseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::WellKnown(e) => e.fmt(f),
+            // Don't display unknowns to end users, only debug logs
+            Self::Unrecognized(_, _) => write!(f, "The receiver sent an unrecognized error."),
+            Self::Validation(e) => write!(f, "The receiver sent an invalid response: {}", e),
+        }
+    }
+}
+
+impl fmt::Debug for ResponseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::WellKnown(e) => write!(
+                f,
+                r#"Well known error: {{ "errorCode": "{}",
+                "message": "{}" }}"#,
+                e.error_code(),
+                e.message()
+            ),
+            Self::Unrecognized(code, msg) => write!(
+                f,
+                r#"Unrecognized error: {{ "errorCode": "{}", "message": "{}" }}"#,
+                code, msg
+            ),
+            Self::Validation(e) => write!(f, "Validation({:?})", e),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum WellKnownError {
+    Unavailable(String),
+    NotEnoughMoney(String),
+    VersionUnsupported(String),
+    OriginalPsbtRejected(String),
+}
+
+impl WellKnownError {
+    pub fn error_code(&self) -> &str {
+        match self {
+            WellKnownError::Unavailable(_) => "unavailable",
+            WellKnownError::NotEnoughMoney(_) => "not-enough-money",
+            WellKnownError::VersionUnsupported(_) => "version-unsupported",
+            WellKnownError::OriginalPsbtRejected(_) => "original-psbt-rejected",
+        }
+    }
+    pub fn message(&self) -> &str {
+        match self {
+            WellKnownError::Unavailable(m) => m,
+            WellKnownError::NotEnoughMoney(m) => m,
+            WellKnownError::VersionUnsupported(m) => m,
+            WellKnownError::OriginalPsbtRejected(m) => m,
+        }
+    }
+}
+
+impl WellKnownError {
+    fn from_error_code(s: &str, message: String) -> Result<Self, ()> {
```

Comment:
Seems like this could shortcut directly to an unrecognized error. Just a note in case this gets refactored in the future.